### ISSUE TITLE
kaniko build never finishes

### ISSUE THREAD

**Issue by daddz** [GoogleContainerTools/skaffold#2529]
### Expected behavior
Kaniko build finishes after uploading to registry.
### Actual behavior
The process never finishes after uploading the image to the registry.
I can see that the kaniko pod finished successfully but skaffold doesn't do anything after that.
The image is also successfully uploaded to the registry.

Tried with both `skaffold build -q > build_test.out` and `skaffold -v info --file-output build_test.out` and there's no info that shows any problem.

### Information

- Skaffold version: v0.34.0
- Operating system: Linux
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v1beta13
kind: Config
build:
  cluster: 
    namespace: kaniko-build
    # needed as workaround until https://github.com/GoogleContainerTools/skaffold/issues/2190 is fixed and released
    pullSecretName: docker-cfg
    # end workaround
    dockerConfig:
      secretName: docker-cfg
  artifacts:
  - image: my-registry/test-image
    kaniko:
      buildContext:
        localDir: {}
      cache:
        repo: my-registry/test-image/cache
      reproducible: true
```

### Steps to reproduce the behavior

1. `skaffold build` with above config

**Comment by balopat** [comment]
Thank you for filing this! Can you please attach logs from the `skaffold build`/dev output, maybe with `-v debug` so that we can see where the process hangs? Thank you!

**Comment by daddz** [comment]
Sure thing, I didn't see there's a debug log level while looking at the CLI reference.

This is the end of the debug output where it's not progressing further:
```
INFO[0154] Taking snapshot of files...                  
DEBU[0154] Taking snapshot of files [/app /app/Dockerfile ... (shortened)]
INFO[0154] WORKDIR /app                                 
INFO[0154] cmd: workdir                                 
INFO[0154] Changed working directory to /app            
INFO[0154] No files changed in this command, skipping snapshotting. 
INFO[0154] ENTRYPOINT ["/scripts/dev"]                  
INFO[0154] No files changed in this command, skipping snapshotting. 
2019/07/24 17:39:18 pushed blob: sha256:1c83652813bc52440ee69db9ba203968ddbf5dddd5379a99f4323550791bd3d7
2019/07/24 17:39:18 pushed blob: sha256:e9ef49d6ae6f09918008083af15ae767f983f7df67c64e2036aef9ea9778fd8b
2019/07/24 17:39:19 pushed blob: sha256:e6406aec7e3075b5bfbbe906c2e5930d60ba44ff24b9de70d2008202dff496fb
2019/07/24 17:39:19 pushed blob: sha256:d3de072f21349d83878f487b93d03ef55e0c8037fc66292f06b07847211f816e
2019/07/24 17:39:21 my-registry/test-image/cache:9658e1d91ed3f48cd1b027644b1f2d1b96188760fd080f675a9136d67a9f24a3: digest: sha256:e957601f9c623e6e70116fa6fe24bcd47a960103bc8964200a41f1f35a7998b1 size: 424
2019/07/24 17:39:21 my-registry/test-image/cache:31333bdea0b54982ce9608eb181a7f8e74fbc1796e1af4c7cdbcecde78f84143: digest: sha256:5183fde5e6946fc0ee199c63ee99e077fee7e8ffcf80e3844f5359b327fd2d8e size: 424
2019/07/24 17:39:21 pushed blob: sha256:2ae45a7a5d93e09df4d0119da9abf9e2ec6af2656200f13fe9aaf1227bd17ebc
2019/07/24 17:40:03 pushed blob: sha256:df9088388eb7e0a3acaa4bf7212641f542e4b19b24ffb1c5b0560b98b442fbcf
2019/07/24 17:40:03 my-registry/test-image/cache:1ecad4372ea6e562771713dd140ac70fbeb7c1429fa55e28a1281821b3b62a9c: digest: sha256:ab3c50d73b3e010d293a51e0e8520429c3cad732124c7c20a604e866a4f54bb7 size: 429
2019/07/24 17:40:30 existing blob: sha256:71275273b8558efe07e2fe71f4fe67e109170e6337b4e4d2df7ba17fe6db26b8
2019/07/24 17:40:30 existing blob: sha256:6ed04e10652ddd56a28f6f8616ecd22914235a06032b9a7b818568b8ba2325c9
2019/07/24 17:40:30 existing blob: sha256:d70f614d461ca9baf55a1f8fea5bfe224d29a45f1e432e5a948bae380b4916ea
2019/07/24 17:40:30 existing blob: sha256:f87e7e672a9a1a6b44ac1b90d30e1f83bc2d304cabfe0325a1b1ba12b5c96541
2019/07/24 17:40:30 existing blob: sha256:dc8825ff62296ef798addcb222026742f55be37fa27c396a8934d33d9189816c
2019/07/24 17:40:31 pushed blob: sha256:c833b680e8dbba4a2b10f3c92fd9fa3bbb8222142c5735e7cf83da4e4f4d57e9
2019/07/24 17:40:32 pushed blob: sha256:08212a5b3797d45b42f5e8daeec848ea0b7929aa01d63c215c7648d1fb3a2788
2019/07/24 17:40:32 pushed blob: sha256:984542ab2baf2bca1da3ebd55206cb2e3caf019cd0e11274386cdb2f377bb245
2019/07/24 17:40:32 pushed blob: sha256:0bfa19551d63859da6f94c0b89132a7741ee8b30996d4178d4e749cf0a306e1b
2019/07/24 17:40:32 pushed blob: sha256:ff35e525113de28c91bbc6a83c2a32961891558f972af985473ce14a1908cbaf
2019/07/24 17:40:46 pushed blob: sha256:b2ff6fe1e41897f568f5cf7190cb049a5400ec0a3ef873fda5ccb4c56a5c282a
2019/07/24 17:40:49 pushed blob: sha256:1d86801a6f8631fa8734f2b95f720c337bd2c312c2f1012b6def0db4fad4e5e9
2019/07/24 17:40:51 pushed blob: sha256:1b545a8c59a0bb2994f75a76ab0a168d7cbcd355e62900e13dfd13b2671e0e4a
2019/07/24 17:41:48 pushed blob: sha256:d57d6fa1c69779ec0e90135252879df1d7e6f7755edd6b10931332d0085e37f6
2019/07/24 17:41:49 my-registry/test-image:8d36003: digest: sha256:97e8b1347e8205a7fbff52906ec0bef08f54ed4bd64fd139750c073f757a343f size: 2381
```

`kubectl` displays the following:
```
➜  ~ kubectl get pods -n kaniko-build
NAME                                                 READY   STATUS      RESTARTS   AGE
kaniko-wsbx4                                         0/1     Completed   0          12m
```
The target registry is a private registry based on Harbor v1.8.0-25bb24ca secured with SSL.
I can also pull the built image without issues and it works as expected.

**Comment by daddz** [comment]
I tried to find the problem myself by digging into the source. Please note that I have no clue about Go so I might be wrong.

File: `skaffold/build/cluster/kaniko.go`
After commenting out the `waitForLogs()` call the whole thing runs through without a problem:
```
	if err := s.ModifyPod(ctx, pod); err != nil {
		return "", errors.Wrap(err, "modifying kaniko pod")
	}

	//waitForLogs := streamLogs(out, pod.Name, pods)

	err = kubernetes.WaitForPodSucceeded(ctx, pods, pod.Name, b.timeout)
	//waitForLogs()
	if err != nil {
		return "", errors.Wrap(err, "waiting for pod to complete")
	}

	return docker.RemoteDigest(tag, b.insecureRegistries)
```

Log output:
```
INFO[0004] Waiting for kaniko-tc927 to be initialized   
DEBU[0007] Running command: [kubectl exec -i kaniko-tc927 -c kaniko-init-container -n kaniko-build -- tar -xzf - -C /kaniko/buildcontext] 
DEBU[0013] Running command: [kubectl exec kaniko-tc927 -c kaniko-init-container -n kaniko-build -- touch /tmp/complete] 
INFO[0017] Waiting for kaniko-tc927 to be complete      
Build complete in 44.289376262s
Starting test...
Test complete in 14.853µs

**Comment by dgageot** [comment]
I spent some time trying to reproduce and failed.

**Comment by daddz** [comment]
Thanks for looking into this!
I just tried to dig into this a little more and so it happens that if you let it sit long enough it eventually finishes:
```
Build complete in 31m2.109745064s
Starting test...
Test complete in 3.33µs
```
I also tried to reproduce it and it always ends up around 30 minutes. Is there a timeout?

After fiddling around in the source this line seems to cause the delay:
https://github.com/GoogleContainerTools/skaffold/blob/7a0f5f1d724316f840877a11a4e02963627e7483/pkg/skaffold/build/cluster/logs.go#L60

If I comment out the related parts like this:
```go
for atomic.LoadInt32(&retry) == 1 {
    _, err := pods.GetLogs(name, &v1.PodLogOptions{
	 Follow:    true,
	 Container: constants.DefaultKanikoContainerName,
    }).Stream()
    if err != nil {
        logrus.Debugln("unable to get kaniko pod logs:", err)
	time.Sleep(1 * time.Second)
	continue
    }
    // w, _ := io.Copy(out, r)
    // atomic.AddInt64(&written, w)
    return
}
```
...it works fine:
```
Build complete in 58.396420204s
Starting test...
Test complete in 17.005µs
```

**Comment by daddz** [comment]
The error seems to be related to the "Follow: true" option.
Stream seems like it never (only after ~30 minutes) gets closed after the pod has succeeded.
I hope this information helps.

**Comment by daddz** [comment]
Some more infos of last test setup:
```
$> kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.2", GitCommit:"f6278300bebbb750328ac16ee6dd3aa7d3549568", GitTreeState:"archive", BuildDate:"2019-08-29T18:43:18Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.2", GitCommit:"f6278300bebbb750328ac16ee6dd3aa7d3549568", GitTreeState:"clean", BuildDate:"2019-08-05T09:15:22Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```
```
$> skaffold version
v0.37.0-81-g7a0f5f1d
```

**Comment by dgageot** [comment]
Oh, `v1.15.2`... This could be the issue. We use kubernetes client 1.12 to retrieve the logs. There might be something broken. I'll see what we can do

**Comment by daddz** [comment]
Ah I haven't thought about that. Glad to hear that you might have a lead.
FYI: when I first started running into this problem it was 1.14.x

**Comment by dgageot** [comment]
@daddz Do you have a way to test https://github.com/GoogleContainerTools/skaffold/pull/2780 and see if it's better?

**Comment by daddz** [comment]
@dgageot I have just tested it and unfortunately it's still the same problem.

It seems to be a problem in kubectl as well.
I replicated the command and ran both in parallel and they both hang and don't return at the same spot.
1. I started: `skaffold build`
2. While build was running I retried `kubectl logs -n default -f kaniko-j8d4j -c kaniko` until it worked (pod ready)

**Comment by dgageot** [comment]
I’d love to be able to test the same kaniko build as you. Because now, I think it might be something on that side.

**Comment by daddz** [comment]
@dgageot thanks for all the effort. You tipped me in the a new direction. The whole thing seems to be related to the way these clusters are deployed.
I tested this issue on other clusters and it does not appear. I'll post more details as soon as i have them.

**Comment by daddz** [comment]
I have finally identified the issue:
All of the nodes were configured with logging driver `journald` for docker.
After switching to `json-file` everything worked without problems.

I checked twice to confirm I haven't gotten insane.
I swapped the node to execute kaniko back to using `journald` as log-driver for docker which results in the symptoms described above. Switch back to `json-file` and it's fine.

What a ride...
